File: binary/python3/maximum_depth_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root, depth=1):
        """
        :type root: TreeNode
        :type depth: int
        :rtype: int
        """
        if root is None:
            return 0
        
        answer = depth
        
        if root.left is not None:
            answer = max(answer, self.maxDepth(root.left, depth+1))
        
        if root.right is not None:
            answer = max(answer, self.maxDepth(root.right, depth+1))
        
        return answer

File: binary/cpp/maximum_depth_of_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    if (!root) {
        return 0;
    }

    if (!root.left && !root.right) {
        return 1;
    }
    
    let leftDepth = 0;
    let rightDepth = 0;
    if (root.left) {
        leftDepth = maxDepth(root.left);
    }
    
    if (root.right) {
        rightDepth = maxDepth(root.right);
    }
    
    if (leftDepth > rightDepth) {
        return leftDepth + 1;
    } else {
        return rightDepth + 1;
    }
};
